cuML/test/trustworthiness_test.cu

```diff
@@ -0,0 +1,434 @@
+            0.34028092,-0.2089237,-0.05382041,0.01625874,-0.02080803,-0.12535211,
+            -0.04146428,-1.24533033,0.48944879,0.0578458,0.26708388,-0.90321028,
+            0.35377088,-0.36791429,-0.35382384,-0.52748734,0.42854419,-0.31744713,
+            -0.19174226,-0.39073724,-0.03258846,-0.19978228,-0.36185205,-0.57412046,
+            0.43681973,-0.25414538,-0.12904905,-0.46334973,-0.03123853,-0.11303604,
+            -0.87073672,-0.45441297,0.41825858,-0.25303507,-0.21845073,0.10248682,
+            -0.11045569,-0.10002795,-0.00572806,0.16519061,0.42651513,-1.11417019,
+            -0.83789682,0.02995787,0.16843079,-0.53874511,0.03056994,0.17877036,
+            0.49632853,-1.03276777,-0.74778616,-0.03971953,0.10907949,-0.67385727,
+            -0.9523471,-0.56550741,0.40409449,-0.2703723,-0.10175014,0.13605487,
+            -0.06306008,-0.01768126,-0.4749442,-0.56964815,0.39389887,-0.19248079,
+            -0.04161081,-0.38728487,-0.20341556,-0.12656988,-0.35949609,-0.46137866,
+            0.28798422,-0.06603147,-0.04363992,-0.60343552,-0.23565227,-0.10242701,
+            -0.06792886,0.09689897,0.33259571,-0.98854214,-0.84444433,0.00673901,
+            0.13457057,-0.43145794,-0.51500046,-0.50821936,0.38000089,0.0132636,
+            0.0580942,-0.40157595,-0.11967677,0.02549113,-0.10350953,0.22918226,
+            0.40411913,-1.05619383,-0.71218503,-0.02197581,0.26422262,-0.34765676,
+            0.06601537,0.21712676,0.34723559,-1.20982027,-0.95646334,0.00793948,
+            0.27620381,-0.43475035,-0.67326003,-0.6137197,0.43724492,-0.17666136,
+            -0.06591748,-0.18937394,-0.07400128,-0.06881691,-0.5201112,-0.61088628,
+            0.4225319,-0.18969463,-0.06921366,-0.33993208,-0.06990873,-0.10288513,
+            -0.70659858,-0.56003648,0.46628812,-0.16090363,-0.0185108,-0.1431348,
+            -0.1128775,-0.0078648,-0.02323332,0.04292452,0.39291084,-0.94897962,
+            -0.63863206,-0.16546988,0.23698957,-0.30633628
+        };
+
+        ML::cumlHandle h;
+        cudaStream_t stream = h.getStream();
+        auto d_alloc = h.getDeviceAllocator();
+
+        float* d_X = (float*)d_alloc->allocate(X.size() * sizeof(float), stream);
+        float* d_X_embedded = (float*)d_alloc->allocate(X_embedded.size() * sizeof(float), stream);
+
+        updateDevice(d_X, X.data(), X.size(), stream);
+        updateDevice(d_X_embedded, X_embedded.data(), X_embedded.size(), stream);
+
+        // euclidean test
+        score = trustworthiness_score<float>(h, d_X, d_X_embedded, 50, 30, 8, 5, 5);
+
+        d_alloc->deallocate(d_X, X.size() * sizeof(float), stream);
+        d_alloc->deallocate(d_X_embedded, X_embedded.size() * sizeof(float), stream);
+	}
+
+	void SetUp() override {
+		basicTest();
+	}
+
+	void TearDown() override {
+	}
+
+protected:
+	double score;
+
+};
+
+
+typedef TrustworthinessScoreTest TrustworthinessScoreTestF;
+TEST_F(TrustworthinessScoreTestF, Result) {
+    ASSERT_TRUE(0.9374 < score && score < 0.9376);
```

Comment:
The knn is implemented as a brute-force matrix multiply, which _should_ make the results deterministic, if I'm not mistaken. Should this just be `ASSERT_TRUE(score == 0.9375)`?

Reply:
Unfortunately, the result doesn't match exactly with sklearn and I don't want the test to fail when that might get fixed. I did a little investigation to pinpoint the problem. It turns out that when using single precision, you have very high probability of pairwise distance collision even for small matrices. Because of that, the sorting step has several potentially valid results. This, of course, affects the final result.

Reply:
When I say "very high probability", it depends a lot on the way the dataset was built. At least using the sklearn's make_blobs function, there's a clear difference in term of distance collisions between single and double precision.

Reply:
This does make sense, actually. This is also not the first time we have encountered problems with single precision inputs in our algorithms. Thank you for being thorough in your investigation!